12-hash__Issue #132__Mohamed_A

Mohamed6886 opened the pull request:
fixes #132 

Added new section on hashability, mutability, and immutability
Fixed syntax for string comparisons in code snippets.

onzfonz reviewed (changes requested):
make sure to add the reference to the java documentation, and introduce a link.

Mohamed6886 commented:
> make sure to add the reference to the java documentation, and introduce a link.

Done, added two resources one on Java Objects that explains hasCode() and equals(), and one on immutability, but they are not showing on the footnotes for some reason.

onzfonz commented on docs/labs/12-Hash.md:
````diff
@@ -87,6 +87,18 @@ the value will be
 
+## How do HashMaps handle mutable vs. immutable objects?
+
+In Java, any object you use as a key for a ```HashMap``` needs to have an ```equals()``` and
+```hashCode()``` that match up correctly--this is called being “hashable.” Immutable objects
+(like ```Strings```) are usually better as keys because they don’t change once created, so
+their ```equals()``` and ```hashCode()``` never change either. If you use objects that can
+change (mutable) like an ```ArrayList```, make sure not to modify the parts that affect
+```equals()``` or ```hashCode()```, or the ```HashMap``` might not find your key anymore!
````
```suggestion
In Java, any object you use as a key for a ```HashMap``` needs to have an ```equals()```
and ```hashCode()``` that match up correctly--this is called being **hashable**.
Immutable objects (like ```Strings```)
are usually better as keys because they don’t change once created,
so their ```equals()``` and ```hashCode()``` never change either.
If you use objects that can change (*mutable*) like an ```ArrayList```,
make sure not to modify the parts that affect ```equals()``` or ```hashCode()```,
or the ```HashMap``` might not find your key anymore!
```

onzfonz commented on docs/labs/12-Hash.md:
````diff
@@ -87,6 +87,18 @@ the value will be
+change (mutable) like an ```ArrayList```, make sure not to modify the parts that affect
+```equals()``` or ```hashCode()```, or the ```HashMap``` might not find your key anymore!
+
+So keep in mind: no matter if your object is immutable or mutable, you must override
+```equals()``` and ```hashCode()``` in a consistent way.
````
```suggestion
So keep in mind: no matter if your object is immutable or mutable,
***you must override ```equals()``` and ```hashCode()``` in a consistent way***.
```

onzfonz reviewed (changes requested):
Make sure you use the pull request review window to add the suggestion to batch and commit them directly from within github.
You'll want to do this correctly and from within the files changed tab of the pull request.
The last thing I'll ask for is to insert the links you provided as footnotes in the text you added.

onzfonz commented:
you'll also need to update your branch.  Feel free to message me as well.

Mohamed6886 commented:
> you'll also need to update your branch. Feel free to message me as well.

Done
- I've added the suggestions to batch and committed them through the "Files changed" tab
- I inserted the links as footnotes by referencing them in the text
- I updated the branch

Let me know if there's anything else you'd like done.
